Cx Organization

JuanPedroGHM opened the pull request:
## Due Diligence
<!--- Please address the following points before setting your PR "ready for review".
--->
- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [ ] unit tests: all split configurations tested
    - [ ] unit tests: multiple dtypes tested
    - [ ] **NEW** unit tests: MPS tested (1 MPI process, 1 GPU)
    - [ ] benchmarks: created for new functionality
    - [ ] benchmarks: performance improved or maintained
    - [ ] documentation updated where needed

## Description

<!--- Include a summary of the change/s.
Please also include relevant motivation and context. List any dependencies that are required for this change.
--->

Goal is to organize our workflows by modularizing certain steps and jobs.
Reusable workflows should be turned into actions, that can be added onto larger workflows. 

Anger towards #1937 made me finish this out of spite. It does all the changes proposed. Let's talk about what's missing.

Benchmarks are also working again.

## What are we getting from this

- More detailed test reports in gitlab and codecov (allegedly)
- Actions that are easier to maintain and workflows that are easier to plan.

**TODO**

- [ ] Clean-up secrets and old tokens
- [x] Change GH_TOKEN, CB cannot comment on pr's at the moment.
- [x] Restore full CI runs (after pr approved, and on main?)
- [ ] Somehow test the weekly workflow.
- [x] Make sure special cases CI cases (support) are somehow handled properly.  

See #1874 Cx section for rationale.

## Changes proposed:

- [x] Weekly actions packed into weekly workflow
- [x] A single Github CI Action
- [x] A single Codebase CI Action
- [x] A single Benchmark trigger action
- [x] A PR open/update workflow
- [x] A PR approve workflow
- [x] A push to main workflow
- [x] NO MORE THANKING YOU FOR THE PR!

JuanPedroGHM committed:
wip: workflows and actions refactoring

brownbaerchen commented:
What is the status of this PR? Is this still relevant after #2139? Some things like removing thanking for PR are certainly still relevant but may be better merged separately.

JuanPedroGHM committed:
wip: we continue

JuanPedroGHM committed:
fix: need to checkout stuff

JuanPedroGHM committed:
fix: did not save

JuanPedroGHM committed:
fix: correct actio definition

JuanPedroGHM committed:
fix: typos

JuanPedroGHM committed:
fix: secret handling in actions

JuanPedroGHM committed:
fix: pre-execution bug for trigger steps

JuanPedroGHM committed:
fix: larger single step

JuanPedroGHM committed:
fix: simplified tests

JuanPedroGHM committed:
fix: simplified trigger actions

JuanPedroGHM committed:
fix: split pr update action

JuanPedroGHM committed:
fix: forgot to checkout individual jobs

JuanPedroGHM committed:
fix: figuring out how this actually works

JuanPedroGHM committed:
fix: removed ready for review

brownbaerchen commented on .github/actions/gh_tests/action.yml:
```diff
@@ -0,0 +1,69 @@
+  install-options:
+    description: "Package install target/options for heat."
+    required: false
+    default: ".[hdf5,netcdf,pandas,zarr]"
```
After merging #2262, we need to do the following (not now!)
```suggestion
    default: ".[dev]"
```

brownbaerchen commented on .github/actions/gh_tests/action.yml:
```diff
@@ -0,0 +1,69 @@
+        architecture: x64
+
+    - name: Install heat
+      if: ${{ inputs.mode == 'full' }}
```
```suggestion
```
I think we can run the few additional IO tests also in the quick checks

brownbaerchen commented on .github/actions/gh_tests/action.yml:
```diff
@@ -0,0 +1,69 @@
+        mpirun -n 1 pytest
+        mpirun -n 2 pytest
+        mpirun -n 3 pytest
+        mpirun -n 4 pytest
```
```suggestion
    - name: Run serial tests
       shell: bash
       run: |
          pytest 
    
    - name: Run tests with 2 tasks    
      if: ${{ inputs.mode == 'full' }}
      shell: bash
      run: |
        mpirun -n 2 pytest
   
     - name: Run tests with 3 tasks    
      if: ${{ inputs.mode == 'full' }}
      shell: bash
      run: | 
        mpirun -n 3 pytest
        
     - name: Run tests with 4 tasks    
      shell: bash
      run: |
           mpirun -n 4 pytest
```

A lot of times we skip heat implementations and pass directly to pytorch if we run in serial, so also for quick checks we should have serial and parallel. 
Also, I like to split this up so you can see more quickly in the action log what exactly failed. If only the parallel tests failed, you can see quickly that the serial ones passed, for instance.

brownbaerchen commented on .github/actions/gh_tests/action.yml:
```diff
@@ -0,0 +1,69 @@
+      if: ${{ inputs.mode == 'quick' }}
+      shell: bash
+      run: |
+        pytest
```
```suggestion
```
This is no longer needed if you accept the above comments

brownbaerchen commented on .github/actions/gh_tests/action.yml:
```diff
@@ -0,0 +1,69 @@
+      shell: bash
+      run: |
+        pip install pytest
+        pip install ${{ inputs.pytorch-version }} ${{ inputs.install-options }} --extra-index-url https://download.pytorch.org/whl/cpu
```
Note that this will then only check with the additional optional dependencies installed, which I think is fine. Actually, I have suggested this myself in #2196, but @mtar wanted to keep a run with no optional dependencies. I am indifferent as long as they run concurrently.

JuanPedroGHM committed:
fix: more test configurations

JuanPedroGHM committed:
--amend

brownbaerchen reviewed (changes requested):
Thanks for cleaning up all this stuff!

 I left a few comments, but I have one larger issue outside of that, which is that no tests currently run on push to non-main branches, right?
I really like running the CI on push to any branch because I don't always run the entire test suite locally. Most of the time, I just run the tests I am working on and I know that I will get an email if I messed up something else. Or maybe the tests pass on my machine and not on GitHub or whatever. Or there is a new optional dependency that I have yet to install. Or maybe I just forgot to check different number of tasks.
I think it's really nice to be able to run the CI on push on my fork before doing a PR because then I can fix failures without sending out a bunch of emails to everybody.

Unfortunately, this is a bit tricky to implement in this framework because we don't want duplicate runs when pushing to a branch within the main repository that is currently part of a PR. There needs to be some check that the CI on push is skipped if the repository is not a fork and if the branch is currently part of a PR or something like that.

What's your opinion on this?

JuanPedroGHM committed:
fix: uv - no system python

JuanPedroGHM committed:
fix: split python and uv action

JuanPedroGHM committed:
fix: addressing review comments

JuanPedroGHM commented:
Hi @brownbaerchen, I forgot how quickly you answer reviews. There were some changes that needed to be tested. But I think the current version now addresses most of the issues you mentioned.

- `push_other.yml` is meant to run only on branches without a PR or forks. 
- Dependencies have been changed to `.[dev]`
- Restored the full CI matrix for pushes to main and stable.
- Using `uv` for caching and faster dependency installation.
- Restored the `push_support` action for testing newer pytorch versions.

JuanPedroGHM committed:
Merge branch 'main' into cx/workflow_action_refactor

brownbaerchen commented on .github/workflows/pr_update.yml:
```diff
@@ -0,0 +1,83 @@
+      - uses: ./.github/actions/gh_tests
+        with:
+          mode: quick
+          sha: ${{ github.event.pull_request.head.sha }}
```
```suggestion
    name: Medium Tests
    runs-on: ubuntu-latest
    steps:
      - name: Harden the runner
        uses: step-security/harden-runner@6c3c2f2c1c457b00c10c4848d6f5491db3b629df # v2.18.0
        with:
          egress-policy: audit
      - name: Checkout code
        uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
        with:
          ref: ${{ github.event.pull_request.head.sha }}
      - uses: ./.github/actions/gh_tests
        with:
          mode: medium
          sha: ${{ github.event.pull_request.head.sha }}
```
I vote still running with multiple tasks here. I don't really see why not since it's pretty fast and having this output on GitHub is convenient.

JuanPedroGHM replied:
So, the point of the quick ones is that they are a one-to-one replacement of the tests from `ReceivePR.yml`, which acted as a quick gatekeeper for the codebase tests. I thought it would be fine to keep it the same way, as we ultimately will care more about the codebase tests, which will run in parallel. But we can have a broader discussion on the next PR talk, as a lot of things changed in the codebase backend as well.

brownbaerchen replied:
Ah, right. Having such a gate keeper is a good idea for sure.
 - quick: Run with one single environment and 1 task, runs before launching anything on codebase
 - medium: Run with three representative environments and 1 and 4 tasks, runs in PRs and on push to development branches
 - full: Run with a bunch of environments and 1-4 tasks, runs on push to main and stable
Right?

brownbaerchen reviewed (dismissed):
This is great, thanks! 

I don't really see the point in the quick CI, to be honest. To me, the current lean CI is fast enough. So I would stick to the current medium and full, while renaming medium to quick just to keep the workflows simpler.
But I am also fine with what the state is now. You (or any other reviewer) choose.

brownbaerchen commented:
We need to merge #2262 before this one in order to run the optional IO tests. That's probably also why coverage is reduced currently..

JuanPedroGHM committed:
Merge branch 'main' into cx/workflow_action_refactor

JuanPedroGHM committed:
fix: corrected support workflow branch

JuanPedroGHM committed:
Merge branch 'main' into cx/workflow_action_refactor

JuanPedroGHM commented:
I fixed the issues with the access token, wrote more documentation on the CX repository, and added the necessary setting for the support new pytorch version workflow. The last pipeline ran without any issue. Ready to merge in my opinion, @brownbaerchen @mtar.

brownbaerchen commented on .github/workflows/push_main.yml:
```diff
@@ -9,10 +8,31 @@ on:
       - 'heat/**'
       - 'tests/**'
       - 'pyproject.toml'
```
One issue that came up is the update of the setup-mpi action. The dependabot PR doesn't trigger any CI action in the current main. Luckily, we know there are issues and so I ran the CI on this manually and indeed it failed. had we not known that, I would want to know at least when the PR is merged into main. So maybe always run the CI on push to main? Not sure what's the best way of handling this.. 
Maybe we should bring back the medium CI on dependabot PRs. The quick one would pass with the setup-MPI thing.

The suggestion diff is messed up. I want to remove all paths, not just the last one.

```suggestion
```

JuanPedroGHM replied:
This would get tested anyway on any push to main, which should be enough for most dependabot updates. For the special cases, like the `setup-mpi` actions, triggering manually is fairly simple with a small commit.

Another option would be to just enable the quick tests when a PR is open, regardless of the source.

brownbaerchen replied:
Is there a reason to only run the CI on push to main with changes to select paths?

brownbaerchen commented on .github/workflows/push_main.yml:
```diff
@@ -9,10 +8,31 @@ on:
+        uses: step-security/harden-runner@6c3c2f2c1c457b00c10c4848d6f5491db3b629df # v2.18.0
+        with:
+          egress-policy: audit
+      - name: Checkout code
```
Do we want to run the quick CI here before triggering stuff on codebase as well? Not sure.

JuanPedroGHM replied:
I think it makes less sense to do it here because 1) there are not that many pushes to main, and we should always test them on all the hardware regardless of whether the simple CI fails or not. 2) The large matrix test would trigger at the same time.

brownbaerchen commented on .github/workflows/push_other.yml:
```diff
@@ -0,0 +1,65 @@
+  push:
+    branches-ignore:
+      - main
+      - dependabot/**
```
```suggestion
```
I don't think we need this since dependabot will always do a PR, I guess?

brownbaerchen commented on .github/workflows/push_support.yml:
```diff
@@ -0,0 +1,23 @@
+name: On Push Main
```
```suggestion
name: On Push Support
```

brownbaerchen commented on .github/workflows/push_support.yml:
```diff
@@ -0,0 +1,23 @@
+      - 'support/**'
+jobs:
+  codebase-test-and-bench-main:
+    name: Testing and
```
```suggestion
    name: Testing and benchmarks
```

brownbaerchen commented on .github/workflows/pr_update.yml:
We also want to trigger the medium CI here, right?

This would also happen for dependabot PRs, which I think is desired. Or maybe we can do a compromise where we trigger the medium CI on PR open for any branch and exclude dependabot branches for PR updates?

Also, I guess we don't want to run the codebase CI on dependabot PRs in general, right?

brownbaerchen commented on .github/workflows/pr_open.yml:
We also need to trigger the medium CI and codebase CI here, no?

brownbaerchen reviewed (changes requested):
I actually noticed a few things I hadn't before:
 - We should trigger more CI on PR open and synchronize
 - We may want to limit CI runs on dependabot PRs, but not as much as we are currently doing

Then again, I may be misunderstanding some stuff. So feel free to dismiss the comments if they don't make sense.

mtar commented on .github/actions/gh_tests/action.yml:
```diff
@@ -0,0 +1,80 @@
+      with:
+        mpi: ${{ inputs.mpi }}
+    - name: "Set up Python"
+      uses: actions/setup-python@v6
```
```suggestion
      uses: actions/setup-python@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.2.0
```

mtar replied:
```suggestion
      uses: actions/setup-python@a309ff8b426b58ec0e2a45f0f869d46889d02405 # v6.2.0
```

mtar commented on .github/workflows/push_main.yml:
```diff
@@ -9,10 +8,31 @@ on:
+        with:
+          branch: "main"
+          sha: ${{ github.sha}}
+          author: ${{ 'heat_team' }}
```
```suggestion
          author: 'heat_team'
```

JuanPedroGHM committed:
fix: review comments

JuanPedroGHM commented:
Updates based on the review comments:

 - On Push PR also triggers when opening/reopening a PR, to get instant tests.
 - On Push PR, the github tests are closer to the medium tests, testing the oldest and newest python and pytorch versions, without any dependencies. It makes sense to do this, we would rather not find the compatibility issues after a merge to main.
 - Tests will still not run dependabot PR's, I don't think it makes sense 99% of the time, but they can be triggered manually by creating a "fake" commit if we feel the need to test it.

mtar commented on .github/workflows/push_other.yml:
```diff
@@ -0,0 +1,64 @@
+    steps:
+      - name: Check if branch has an open PR
+        id: check-pr
+        uses: actions/github-script@v7
```
```suggestion
        uses: actions/github-script@3a2844b7e9c422d3c10d287c895573f7108da1b3 # v9.0.0
```

mtar commented on .github/actions/gh_tests/action.yml:
```diff
@@ -0,0 +1,80 @@
+      with:
+        python-version: "${{ inputs.py-version }}"
+    - name: Install uv and set the Python version
+      uses: astral-sh/setup-uv@v7
```
```suggestion
      uses: astral-sh/setup-uv@08807647e7069bb48b6ef5acd8ec9567f424441b # v8.1.0
```

mtar reviewed (commented):
Some action SHA pinning for security. It's non-critical code though.

JuanPedroGHM committed:
Apply suggestions from code review

Co-authored-by: Michael Tarnawa <18899420+mtar@users.noreply.github.com>

JuanPedroGHM commented:
@brownbaerchen From your last comment: changes to github workflows will be tested on PR Open/Update (unless they have `dependabot/**` branch), to ensure that they work (just like the ones here), but I would say, once they work on the PR, there is no need to test them on main. Many dependabot merges won't clog the CI anymore. Same principle applies to documentation changes, or other non-source code changes.

JuanPedroGHM committed:
Merge branch 'main' into cx/workflow_action_refactor

JuanPedroGHM committed:
fix: changed openmpi for mpich for gh tests

JuanPedroGHM committed:
fix: removed empty folder

JuanPedroGHM commented:
# Benchmarks results - Sponsored by [perun](https://github.com/Helmholtz-AI-Energy/perun)

## FTPx86-AMD-Milan-Mi250 - CPU - N4 - RUNTIME

| Function   | Value [s]   | Change %   | 30 day avg. [s]   | Last entrie [s]   |
|------------|-------------|------------|-------------------|-------------------|
---

## FTPx86-AMD-Milan-Mi250 - GPU - N4 - RUNTIME

| Function   | Value [s]   | Change %   | 30 day avg. [s]   | Last entrie [s]   |
|------------|-------------|------------|-------------------|-------------------|
---

[Grafana Dashboard](https://930000e0-e69a-4939-912e-89a92316b420.ka.bw-cloud-instance.org)
Last updated: 2026-05-20T13:37:26Z

JuanPedroGHM commented:
Pipeline running after downgrading ROCM to 6.4, also changed github actions to use mpich, for the tests.

JuanPedroGHM committed:
Merge branch 'main' into cx/workflow_action_refactor

JuanPedroGHM committed:
fix: missing flaky install

JuanPedroGHM committed:
fix: added flaky tests

brownbaerchen commented on .github/workflows/push_main.yml:
```diff
@@ -9,10 +8,31 @@ on:
 jobs:
-  full_matrix_test:
+  codebase-test-and-bench-main:
+    name: Codebase Tests and Benchmarks
```
```suggestion
    name: Codebase Tests and Benchmarks
    if: ${{ !github.event.repository.fork }}
```

brownbaerchen commented on .github/workflows/weekly.yml:
```diff
@@ -0,0 +1,121 @@
+    pull-requests: write
+jobs:
+    codeql:
+        name: CodeQL Analysis
```
```suggestion
        name: CodeQL Analysis
        if: ${{ !github.event.repository.fork }}
```

brownbaerchen commented on .github/workflows/weekly.yml:
```diff
@@ -0,0 +1,121 @@
+            with:
+                category: "/language:python"
+    scorecard:
+        name: Code Quality Scorecard
```
```suggestion
        name: Code Quality Scorecard
        if: ${{ !github.event.repository.fork }}
```

brownbaerchen commented on .github/workflows/weekly.yml:
```diff
@@ -0,0 +1,121 @@
+            uses: github/codeql-action/upload-sarif@fca7ace96b7d713c7035871441bd52efbe39e27e # v3.28.19
+            with:
+                sarif_file: results.sarif
+    check-links:
```
```suggestion
    check-links:
      if: ${{ !github.event.repository.fork }}
```

brownbaerchen reviewed (commented):
I added a few skip statements to jobs that should be skipped on forks. Otherwise, this looks really good. Once push to main doesn't trigger codebase CI on forks, it's ready to be merged in my opinion.

JuanPedroGHM committed:
Apply suggestions from code review

Co-authored-by: Thomas Saupe <39156931+brownbaerchen@users.noreply.github.com>

brownbaerchen committed:
Fix indentation

brownbaerchen commented:
Thanks, @JuanPedroGHM! This is really a substantial improvement!
